### PR TITLE
optimize unique url for scanning resources from both base and module

### PR DESCRIPTION
**Pull request by lvjing2**
### Motivation:
1. if can't parse artifact Id from url, then just set as unique and declared
2. set default as declared if can't find artifact
3. add search pom.properties from local directory, if not from target/test-classes, target/classes
4. optimize parse artifact in one method, support 7 kinds of url
5. cache the artifact parse result for local class path.

### CODE REVIEW

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java [hunk 1]
```diff
@@ -460,29 +441,20 @@ public boolean isDeclaredMode() {
...
 
     private boolean doCheckDeclared(String jarFilePath) {
-        String artifactId = "";
-        if (jarFilePath.contains(".jar")) {
-            artifactId = JarUtils.getJarArtifactId(jarFilePath);
-            // if in jar, and can't get artifactId from jar file, then just rollback to all delegate.
-            if (artifactId == null) {
+        String artifactId = JarUtils.parseArtifactId(jarFilePath);
+        if (artifactId == null) {
+            if (jarFilePath.contains(".jar!")) {
```
[the remark this answers is not in the document]
Comment (lvjing2) [reply sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java]: set all as declared, so don't need to check this anymore

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java [hunk 2]
```diff
@@ -460,29 +441,20 @@ public boolean isDeclaredMode() {
...
                     jarFilePath));
                 return true;
-            }
-        } else {
-            artifactId = getArtifactIdFromLocalClassPath(jarFilePath);
-            // for not in jar, then default not delegate.
-            if (artifactId == null) {
+            } else {
+                // for not in jar, then default not delegate.
                 LOGGER.info(String.format(
```
[the remark this answers is not in the document]
Comment (lvjing2) [reply sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/BizModel.java]: set all as declared

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/classloader/AbstractClasspathClassLoader.java [hunk 3]
```diff
@@ -313,32 +312,21 @@ public Enumeration<URL> getResources(String name) throws IOException {
...
-                    artifactId = JarUtils.getArtifactIdFromLocalClassPath(filePath);
+                if (filePath.endsWith(resourceName)) {
+                    filePath = filePath.substring(0, filePath.lastIndexOf(resourceName));
                 }
-                if (!temp.contains(artifactId)) {
+                String artifactId = JarUtils.parseArtifactId(filePath);
+                if (artifactId == null || !temp.contains(artifactId)) {
                     uniqueUrls.add(resourceUrl);
                 }
                 temp.add(artifactId);
```
[the remark this answers is not in the document]
Comment (lvjing2) [reply sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/classloader/AbstractClasspathClassLoader.java]: change to using map putIfAbsent

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/classloader/AbstractClasspathClassLoader.java [hunk 4]
```diff
@@ -313,38 +312,28 @@ public Enumeration<URL> getResources(String name) throws IOException {
...
-                    String jarFilePath = filePath;
-                    if (index != -1) {
-                        jarFilePath = filePath.substring(0, index);
-                    }
-
-                    artifactId = JarUtils.getJarArtifactId(jarFilePath);
-                } else {
-                    artifactId = JarUtils.getArtifactIdFromLocalClassPath(filePath);
+                if (filePath.endsWith(resourceName)) {
+                    filePath = filePath.substring(0, filePath.lastIndexOf(resourceName));
```
Comment (yuanyuancin) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/classloader/AbstractClasspathClassLoader.java:326]: 改变了Classpath中的顺序，可能有不预期的行为

**Review by yuanyuancin — APPROVED** [review]
LGFM

**Review by yuanyuancin — APPROVED** [review]
LGFM

**Pull request merged**